File: solution/002.Add_Two_Numbers/Solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
	var carry = 0;
	var lastNode = null;
	var headNode = null;
	while (l1 || l2 || carry > 0) {
		var result = carry + (l1 ? l1.val : 0) + (l2 ? l2.val : 0);
		var node = new ListNode(result % 10);
		if (!headNode) {headNode = node;}
		if (lastNode) {lastNode.next = node;}
		lastNode = node;
		carry = Math.floor(result / 10);
		if (l1) {l1 = l1.next;}
		if (l2) {l2 = l2.next;}
	}
	return headNode;
};


// test helper
function ListNode(val) {
    this.val = val;
    this.next = null;
}

ListNode.prototype.toString = function() {
	var p = this;
	var result = [];
	while (p) {
		result.push(p.val);
		p = p.next;
	}
	return result.join(" -> ");
};

function numbersToList() {
	var headNode = new ListNode(0);
	var currNode = headNode;
	for (var i = 0; i < arguments.length; i++) {
		var node = new ListNode(arguments[i]);
		currNode.next = node;
		currNode = node;
	}
	return headNode.next;
}

// test
function test() {
	var node = addTwoNumbers(
		numbersToList(2, 4, 3),
		numbersToList(5, 6, 4)
	);
	console.log('result: ' + node.toString());
}

test();

File: solution/002.Add_Two_Numbers/Solution.py
#!/usr/bin/env python

import math


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

    def append(self, val):
        """
        :type val: int
        """
        cur = self
        while cur.next is not None:
            cur = cur.next
        cur.next = ListNode(val)

    def __str__(self):
        l = [str(self.val)]
        node = self.next
        while node is not None:
            l.append(str(node.val))
            node = node.next
        return " -> ".join(l)


class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        head_node = None
        last_node = None
        while carry != 0 or l1 is not None or l2 is not None:
            result = carry + (l1.val if l1 is not None else 0) + (l2.val if l2 is not None else 0)
            n = ListNode(result % 10)
            if head_node is None:
                head_node = n
            if last_node is not None:
                last_node.next = n
            last_node = n
            carry = math.floor(result / 10)
            if l1 is not None:
                l1 = l1.next
            if l2 is not None:
                l2 = l2.next
        return head_node


if __name__ == '__main__':
    l1 = ListNode(2)
    l1.append(4)
    l1.append(3)

    l2 = ListNode(5)
    l2.append(6)
    l2.append(4)

    l3 = Solution().addTwoNumbers(l1, l2)
    print(l3)
